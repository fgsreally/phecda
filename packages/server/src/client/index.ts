interface PhecdaRequest {
  body: Record<string, any>
  query: Record<string, string>
  params: Record<string, string>
  realParam: string
}
export function toReq(arg: PhecdaRequest) {
  const { body, query, realParam } = arg
  return { body, query: Object.keys(query).length > 0 ? `?${Object.entries(query).map(([k, v]) => `${k}=${v}`).join('&')}` : '', params: realParam }
}

export function mergeReq(arg: PhecdaRequest) {
  const { body, query, params } = arg
  return { body, query, params }
}
