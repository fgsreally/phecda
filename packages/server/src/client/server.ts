import type amqplib from 'amqplib'
export * from './axios'

export function createMqReq(channel: amqplib.Channel): <R>(arg: R) => Promise<void> {
  return async (arg: any) => {
    const { url, body } = arg
    await channel.assertQueue(url)
    await channel.sendToQueue(url, Buffer.from(JSON.stringify(body)))
  }
}
