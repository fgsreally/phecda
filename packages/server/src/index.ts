import 'reflect-metadata'

type Construct<T = any> = new (...args: Array<any>) => T

export function Factory<T>(Module: Construct<T>) {
  const moduleMap = new Map<Construct, InstanceType<Construct>>()
  const app = buildNestModule(Module, moduleMap) as InstanceType<Construct<T>>
  return { app, moduleMap }
}

function buildNestModule(Module: Construct, map: Map<Construct, InstanceType<Construct>>) {
  const paramtypes = getParamtypes(Module) as Construct[]
  let instance: InstanceType<Construct>
  if (map.has(Module)) {
    instance = map.get(Module)
    if (!instance)
      throw new Error(`exist Circular Module dep--${Module}`)
    return instance
  }
  map.set(Module, undefined)
  if (paramtypes) {
    instance = new Module(...paramtypes.map(item =>
      buildNestModule(item, map),
    ))
  }
  else {
    instance = new Module()
  }
  map.set(Module, instance)

  return instance
}

function getParamtypes(Module: Construct) {
  return Reflect.getMetadata('design:paramtypes', Module)
}
