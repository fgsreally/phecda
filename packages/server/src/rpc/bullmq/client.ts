/* eslint-disable no-new */
/* eslint-disable prefer-promise-reject-errors */
import { EventEmitter } from 'events'
import type { ConnectionOptions } from 'bullmq'
import { Queue, Worker } from 'bullmq'
import type { ToClientMap } from '../../types'
import type { RpcClientOptions } from '../helper'
import { genClientQueue } from '../helper'

export async function createClient<S extends Record<string, any>>(connectOpts: ConnectionOptions, controllers: S, opts?: RpcClientOptions) {
  let eventId = 1
  let eventCount = 0

  const ret = {} as ToClientMap<S>
  const emitter = new EventEmitter()

  const clientQueue = genClientQueue(opts?.key)

  const queueMap: Record<string, Queue> = {}

  new Worker(clientQueue, async (job) => {
    const { data, id, error } = job.data
    emitter.emit(id, data, error)
  }, { connection: connectOpts })

  for (const i in controllers) {
    ret[i] = new Proxy(new controllers[i](), {
      get(target, p: string) {
        if (typeof target[p] !== 'function')
          throw new Error(`"${p}" in "${i}" is not an exposed rpc `)

        let { tag, queue, isEvent } = target[p]()

        return async (...args: any) => {
          if (!queue)
            queue = tag
          if (!(queue in queueMap))
            queueMap[queue] = new Queue(queue, { connection: connectOpts })

          const id = `${eventId++}`

          queueMap[queue].add(`${tag}-${p}`, {
            id,
            args,
            tag,
            queue: clientQueue,
            func: p,
          })

          if (isEvent)
            return null

          return new Promise((resolve, reject) => {
            if (opts?.max && eventCount >= opts.max)
              reject({ type: 'exceeded' })

            let isEnd = false
            const timer = setTimeout(() => {
              if (!isEnd) {
                eventCount--
                emitter.off(id, listener)
                reject({ type: 'timeout' })
              }
            }, opts?.timeout || 5000)

            function listener(data: any, error: boolean) {
              eventCount--
              isEnd = true
              clearTimeout(timer)
              if (error)
                reject(data)

              else
                resolve(data)
            }
            eventCount++
            emitter.once(id, listener)
          })
        }
      },
    })
  }

  return ret
}
