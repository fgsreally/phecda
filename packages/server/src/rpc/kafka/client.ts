import { EventEmitter } from 'events'
import { randomUUID } from 'crypto'
import type { Consumer, Producer } from 'kafkajs'
import type { ToClientMap } from '../../types'
export async function createClient<S extends Record<string, any>>(producer: Producer, consumer: Consumer, controllers: S): Promise<ToClientMap<S>> {
  await producer.connect()
  await consumer.connect()

  const ret = {} as any
  const emitter = new EventEmitter()
  const genQueue = (name: string) => `PS:${name}`
  const genReturnQueue = (name: string) => `${name}/return`
  const existQueue = new Set<string>()

  for (const i in controllers) {
    ret[i] = new Proxy(new controllers[i](), {
      get(target, p: string) {
        if (typeof target[p] !== 'function')
          throw new Error(`"${p}" in "${i}" is not an exposed rpc `)

        const { tag, rpc, isEvent } = target[p]()
        if (!rpc.includes('*') && !rpc.includes('kafka'))
          throw new Error(`"${p}" in "${i}" doesn't support kafka`)
        return async (...args: any) => {
          const queue = genQueue(tag)
          const returnQueue = genReturnQueue(queue)
          if (!isEvent && !existQueue.has(returnQueue)) {
            existQueue.add(returnQueue)

            await consumer.subscribe({ topic: queue, fromBeginning: true })
          }
          const id = isEvent ? '' : randomUUID()
          producer.send({
            topic: queue,
            messages: [
              {
                value: JSON.stringify({
                  id,
                  method: p,
                  args,
                }),
              },
            ],
          })

          if (isEvent)
            return null

          return new Promise((resolve, reject) => {
            emitter.once(id, (data: any, error: boolean) => {
              if (error)
                reject(data)

              else resolve(data)
            })
          })
        }
      },
    })
  }
  await consumer.run(
    {
      eachMessage: async ({ message, topic }) => {
        if (!message.value || !existQueue.has(topic))
          return

        const { data, id, error } = JSON.parse(message.value.toString())

        emitter.emit(id, data, error)
      },
    },
  )

  return ret
}
