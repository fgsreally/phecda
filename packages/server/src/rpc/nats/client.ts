/* eslint-disable prefer-promise-reject-errors */
import { StringCodec } from 'nats'
import type { NatsConnection } from 'nats'
import type { ToClientMap } from '../../types'
import type { RpcClientOptions } from '../types'

export async function createClient<S extends Record<string, any>>(nc: NatsConnection, controllers: S, opts?: Omit<RpcClientOptions, 'key'>) {
  let eventId = 1
  let eventCount = 0
  const sc = StringCodec()

  const ret = {} as ToClientMap<S>

  for (const i in controllers) {
    ret[i] = new Proxy(new controllers[i](), {
      get(target, p: string) {
        if (typeof target[p] !== 'function')
          throw new Error(`"${p}" in "${i}" is not an exposed rpc `)

        let { tag, queue, isEvent } = target[p]()

        return async (...args: any) => {
          if (!queue)
            queue = tag

          const id = `${eventId++}`
          const request = nc.request(queue, sc.encode(JSON.stringify({
            _ps: 1,

            id,
            args,
            tag,
            func: p,
          })))

          if (isEvent)
            return null

          return new Promise((resolve, reject) => {
            if (opts?.max && eventCount >= opts.max)
              reject({ type: 'exceeded' })
            // @todo still throw global promise reject
            request.catch(reject)
            request
              .then((msg) => {
                const { data, id, error } = msg.json() as any
                if (id)
                  handler(data, error)
              })

            let isEnd = false
            const timer = setTimeout(() => {
              if (!isEnd) {
                eventCount--
                reject({ type: 'timeout' })
              }
            }, opts?.timeout || 5000)

            function handler(data: any, error: boolean) {
              eventCount--
              isEnd = true
              clearTimeout(timer)
              if (error)
                reject(data)

              else
                resolve(data)
            }
            eventCount++
          })
        }
      },
    })
  }

  return ret
}
