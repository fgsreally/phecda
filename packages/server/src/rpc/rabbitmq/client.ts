import { EventEmitter } from 'events'
import { randomUUID } from 'crypto'
import type amqplib from 'amqplib'
import type { ToClientMap } from '../../types'
import { generateReturnQueue } from '../helper'
export async function createClient<S extends Record<string, any>>(ch: amqplib.Channel, controllers: S, opts?: { timeout?: number }): Promise<ToClientMap<S>> {
  const ret = {} as any
  const emitter = new EventEmitter()

  const existQueue = new Set<string>()

  for (const i in controllers) {
    ret[i] = new Proxy(new controllers[i](), {
      get(target, p: string) {
        if (typeof target[p] !== 'function')
          throw new Error(`"${p}" in "${i}" is not an exposed rpc `)

        let { tag, queue, isEvent } = target[p]()

        return async (...args: any) => {
          if (!queue)
            queue = tag
          const returnQueue = generateReturnQueue(queue)

          if (!isEvent && !existQueue.has(returnQueue)) {
            existQueue.add(returnQueue)
            await ch.assertQueue(returnQueue)
            ch.consume(returnQueue, (msg) => {
              if (!msg)
                return
              const { data, id, error } = JSON.parse(msg.content.toString())
              emitter.emit(id, data, error)
            })
          }
          const id = isEvent ? '' : randomUUID()

          ch.sendToQueue(queue, Buffer.from(
            JSON.stringify(
              {
                id,
                args,
                tag,
                method: p,
              },
            ),
          ))
          if (isEvent)
            return null

          return new Promise((resolve, reject) => {
            const timer = setTimeout(() => {
              emitter.off(id, listener)
              // eslint-disable-next-line prefer-promise-reject-errors
              reject({ message: 'timeout' })
            }, opts?.timeout || 5000)

            function listener(data: any, error: boolean) {
              clearTimeout(timer)
              if (error)
                reject(data)

              else
                resolve(data)
            }
            emitter.once(id, listener)
          })
        }
      },
    })
  }

  return ret
}
