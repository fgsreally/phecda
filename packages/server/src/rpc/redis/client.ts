/* eslint-disable prefer-promise-reject-errors */
import EventEmitter from 'events'
import type Redis from 'ioredis'
import type { ToClientMap } from '../../types'
import type { RpcClientOptions } from '../types'

import { genClientQueue } from '../helper'

export async function createClient<S extends Record<string, any>>({ pub, sub }: { pub: Redis; sub: Redis }, controllers: S, opts?: RpcClientOptions) {
  const ret = {} as ToClientMap<S>
  let eventId = 1
  let eventCount = 0
  const emitter = new EventEmitter()
  const clientQueue = genClientQueue(opts?.key)

  await sub.subscribe(clientQueue)

  for (const i in controllers) {
    ret[i] = new Proxy(new controllers[i](), {
      get(target, p: string) {
        if (typeof target[p] !== 'function')
          throw new Error(`"${p}" in "${i}" is not an exposed rpc `)

        let { tag, queue, isEvent } = target[p]()

        return async (...args: any) => {
          if (!queue)
            queue = tag

          const id = `${eventId++}`

          pub.publish(queue, JSON.stringify({
            _ps: 1,
            args,
            id,
            queue: clientQueue,
            tag,
            func: p,
          }))
          if (isEvent)
            return null

          return new Promise((resolve, reject) => {
            if (opts?.max && eventCount >= opts.max)
              reject({ type: 'exceeded' })

            let isEnd = false
            const timer = setTimeout(() => {
              if (!isEnd) {
                eventCount--
                emitter.off(id, listener)
                reject({ type: 'timeout' })
              }
            }, opts?.timeout || 5000)

            function listener(data: any, error: boolean) {
              eventCount--
              isEnd = true
              clearTimeout(timer)
              if (error)
                reject(data)

              else
                resolve(data)
            }
            eventCount++
            emitter.once(id, listener)
          })
        }
      },
    })
  }

  sub.on('message', async (channel, msg) => {
    if (channel === clientQueue && msg) {
      const { data, id, error } = JSON.parse(msg)

      emitter.emit(id, data, error)
    }
  })

  return ret
}
