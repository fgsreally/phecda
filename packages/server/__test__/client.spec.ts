import { describe, expect, it } from 'vitest'
import { mergeReq, toReq } from '../src/client'

describe('client ', () => {
  it('handle request data', () => {
    const data = {
      query: {
        id: '1',
        name: 'phecda',
      },
      params: {
        test: 'phecda',
        loc: 'loc',
      },
      body: {
        name: 'server',
      },
      realParam: '/phecda/loc',
    }
    expect(toReq(data)).toMatchSnapshot()
    expect(mergeReq(data)).toMatchSnapshot()
  })
})
