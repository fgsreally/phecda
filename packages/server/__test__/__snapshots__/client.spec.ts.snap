// Vitest Snapshot v1

exports[`client  > handle request data 1`] = `
{
  "body": {
    "name": "server",
  },
  "params": "/phecda/loc",
  "query": "?id=1&name=phecda",
}
`;

exports[`client  > handle request data 2`] = `
{
  "body": {
    "name": "server",
  },
  "params": {
    "loc": "loc",
    "test": "phecda",
  },
  "query": {
    "id": "1",
    "name": "phecda",
  },
}
`;
